Deprecate MM_ConfigurationLanguageInterface etc

ktbriggs opened the pull request:
Refactored MM_Configuration and subclasses to clean up convoluted
inheritance factors and eliminate all MM_ConfigurationLangageInterface
usage, replacing with equivalent ConfigurationDelegate method calls.

Client language will be required to implement all methods defined
for MM_ConfigurationDelegate in glue/ConfigurationDelegate. This 
is a more focused articulation of the integration points required for
client languages to configure GC.

Deleted obsolete ConfigurationLanguageInterface*.*.

Updated copyright on all files touched in this commit.

Fixed newlines in ConfigurationGenerational.[ch]pp.

Issue: #508
Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

charliegracie commented:
@ktbriggs quick note before I have reviewed.  Can you limit the first line of the commit message to 50 characters?  Thanks

ktbriggs commented:
Shortened first line of commit message. Also added comments for methods in example/glue/ConfigurationDelegate.hpp and added MM_EnvironmentBase* parameter to two MM_ConfigurationDelegate methods that may require environment to access extensions, vm, port, etc.

ktbriggs commented:
Not sure where error in Travis build is coming from. Travis reports build failure for 32-bit linux only, all other platforms pass. Travis log output for 32-bit linux build is empty. I am going to touch commit message only and push again to force new Travis build.

charliegracie commented on gc/base/Configuration.cpp:
```diff
@@ -162,41 +143,51 @@ MM_EnvironmentBase*
-		return NULL;
+	if (NULL != env) {
+		if (initializeEnvironment(env)) {
+			if (extensions->fvtest_disableInlineAllocation) {
```
It is not clear where this change came from as it did not exist in this function before. It feels like it belongs in the initializeEnvironment method.

charliegracie commented on example/glue/ConfigurationDelegate.hpp:
```diff
@@ -0,0 +1,146 @@
+	 * @return false if heap initialization fails
+	 */
+	bool
+	initializeHeap(MM_EnvironmentBase* env)
```
The name of this function sounds to me like it is initializing the heap.  Maybe we can change it to heapInitialized or something to make it clear that it is called after the heap is initialized to allow the language to do its own initialization of structures, etc. for the heap.

ktbriggs replied:
I agree. This and initializeEnvironment() are both informational (hooks) called from OMR to permit language to respond to GC-related events. Using your suggested name pattern for both of these and will carry this convention forward into ELI and GCLI component delgates.

charliegracie commented on example/glue/ConfigurationDelegate.hpp:
```diff
@@ -0,0 +1,146 @@
+	 * @param env The calling thread environment
+	 * @return A pointer to segregated heap size distribution metadata, or NULL if not using segregated heap
+	 */
+	OMR_SizeClasses *getSegregatedSizeClasses(MM_EnvironmentBase *env)
```
Should this function be wrapped in an #if defined since it only needs to be there if you are using a segregated heap?

ktbriggs replied:
It is unconditionally called from ConfigurationRealtime. All of the realtime code in Java is compiled unconditionally.

charliegracie replied:
Since I have requested another change below can you wrap this code in #if defined()?  As we add new APIs I would like them to all be properly #if defined so it is obvious when you need to worry about them.

ktbriggs replied:
Done. I removed MM_Configuration::initializeSizeClasses(), which previously called this function, and let MM_ConfigurationSegregated and MM_ConfigurationRealtime (Java) call it directly as they are the only consumers of hte segregated heap. The delegate method getSegregatedSizeClasses() is conditionally compiled only when OMR_GC_SEGREGATED_HEAP is enabled.

charliegracie commented on example/glue/ConfigurationDelegate.hpp:
```diff
@@ -0,0 +1,146 @@
+	 */
+	bool
+	initializeEnvironment(MM_EnvironmentBase* env)
+	{
```
Similar to the initializeHeap function this is not actually initializing an environment but rather allowing the language to do its own initialization when an environment is initialized.  Also this code seems like it would be common to all languages and does not belong in the delegate.

ktbriggs replied:
Renaming this function to environmentInitialized(), consistent with heapInitialized(). This (example glue) code is necessary for now but will obsolete when MM_EnvironmentLanguageInterface is refactored into MM_EnvironmentDelegate (WIP). Example code has nothing to add to environment initialization but other languages should have this hook.

charliegracie commented on example/glue/ConfigurationDelegate.hpp:
```diff
@@ -0,0 +1,146 @@
+	 */
+	uintptr_t getMaxGCThreadCount(MM_EnvironmentBase* env)
+	{
+		return 1;
```
Is this for concurrent threads or for slave threads?

ktbriggs replied:
Concurrent threads.
```
void
MM_Configuration::initializeGCThreadCount(MM_EnvironmentBase* env)
{
	...
	extensions->gcThreadCount = omrsysinfo_get_number_CPUs_by_type(OMRPORT_CPU_TARGET);
	if (_delegate.getMaxGCThreadCount(env) < extensions->gcThreadCount) {
		extensions->gcThreadCount = _delegate.getMaxGCThreadCount(env);
	}
	...
```
MM_GCExtensionsBase::gcThreadCount is used to limit threads in MM_ParallelDispatcher:
```
MM_ParallelDispatcher::initialize(MM_EnvironmentBase *env)
{
	...
	_threadCountMaximum = env->getExtensions()->gcThreadCount;
```

charliegracie replied:
That is actually the Slave threads and not the concurrent threads.  Can you update the documentation to reflect this?  Instead of using slave threads you can talk about this number being used to limit the number of threads created by the paralleldispatcher or something like that.

ktbriggs replied:
Done.

charliegracie commented on gc/base/standard/ConfigurationGenerational.hpp:
```diff
@@ -1,6 +1,6 @@
 /*******************************************************************************
  *
- * (c) Copyright IBM Corp. 1991, 2016
+ * (c) Copyright IBM Corp. 1991, 2016, 2017
```
This is not the correct way to update the copyright.  It is initial year, current year.  Can you fix this in all of the files please?

ktbriggs replied:
Ok.

ktbriggs commented:
@charliegracie requested changes are in most recent commit (103396e). All 3 commits have been rebased on eclipse/master@5939598.

ktbriggs commented:
@charliegracie changes reflecting updated from your most recent comments are in commit 4c25835. Will squash these commits when review is complete.
